File: python3/minimum_absolute_difference_in_bst.py
from model import TreeNode

"""
Space   : O(n)
Time    : O(n*edges)
"""

class Solution:
    def getMinimumDifference(self, root: TreeNode) -> int:
        if not root:
            return 0

        ans = 10**10

        stack = [(root, [])]

        while stack:
            node, prev = stack.pop(0)
            if prev:
                for i in prev:
                    ans = min(ans, abs(i - node.val))
            if node.left:
                stack.append((node.left, prev + [node.val]))
            if node.right:
                stack.append((node.right, prev + [node.val]))

        return ans

File: golang/minimum_absolute_difference_in_bst.go
package main

import "golang/model"

/*
Space	: O(n)
Time	: O(n)
*/

func preorderDifInBST(node *model.TreeNode) []int {
	res := []int{}
	if node != nil {
		res = append(res, preorderDifInBST(node.Left)...)
		res = append(res, node.Val)
		res = append(res, preorderDifInBST(node.Right)...)
	}
	return res
}

func getMinimumDifference(root *model.TreeNode) int {
	if root == nil {
		return 0
	}

	ans := 100000
	pre := preorderDifInBST(root)

	if len(pre) == 0 {
		return pre[0]
	}

	for i := 1; i < len(pre); i++ {
		ans = Min(ans, pre[i]-pre[i-1])
	}
	return ans
}
